cmd/query-tee/proxy.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+type Proxy struct {
+	cfg      Config
+	backends []*ProxyBackend
+	logger   log.Logger
+	metrics  *ProxyMetrics
+}
+
+func NewProxy(cfg Config, registerer prometheus.Registerer) (*Proxy, error) {
+	p := &Proxy{
+		cfg:     cfg,
+		logger:  util.Logger,
+		metrics: NewProxyMetrics(registerer),
+	}
+
+	// Parse the backend endpoints (comma separated).
+	parts := strings.Split(cfg.BackendEndpoints, ",")
+
+	for _, part := range parts {
+		u, err := url.Parse(part)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid backend endpoint %s", part)
+		}
+
+		// The backend name is hardcoded as the backend hostname.
+		name := u.Hostname()
+
+		p.backends = append(p.backends, NewProxyBackend(name, u, cfg.BackendReadTimeout, name == cfg.PreferredBackend))
+	}
+
+	// At least 2 backends are required
+	if len(p.backends) < 2 {
```

Comment:
Proxy can also run with single backend, right? I mean, it's pointless, but it's not a hard requirement to have at least two, or is it?

Reply:
It can, but I'm missing the use case. If you configured it with less than 2 backends, it's probably a misconfiguration, reason why I'm reporting it as an error. Thoughts?

Reply:
Possible use case, starting with one backend, making sure it works, then adding another one? Agree it's not a great one, I just feel that warning here would be enough.

Reply:
Ok. Don't have strong opinions. Lowered the requirement to 1 and warning of it's just 1.